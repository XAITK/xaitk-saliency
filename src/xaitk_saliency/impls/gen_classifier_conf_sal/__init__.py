"""Module for all implementations of the GenClassifierConfSal interface"""
