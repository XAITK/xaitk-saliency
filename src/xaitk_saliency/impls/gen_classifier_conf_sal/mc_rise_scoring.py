"""Implementation of MC-RISE scorer"""

from typing import Any

import numpy as np
from sklearn.preprocessing import maxabs_scale
from typing_extensions import override

from xaitk_saliency import GenerateClassifierConfidenceSaliency
from xaitk_saliency.utils.masking import weight_regions_by_scalar


class MCRISEScoring(GenerateClassifierConfidenceSaliency):
    """
    Saliency map generation based on the MC-RISE implementation.
    This version utilizes only the input perturbed image confidence predictions
    and does not utilize reference image confidences.
    This implementation also takes influence from debiased RISE and may take an
    optional debias probability, `p1` (0 by default).
    In the original paper this is paired with the same probability used in RISE
    perturbation mask generation (see the `p1` parameter in
    :class:`xaitk_saliency.impls.perturb_image.mc_rise.MCRISEGrid`).

    Based on Hatakeyama et. al:
    https://openaccess.thecvf.com/content/ACCV2020/papers/Hatakeyama_Visualizing_Color-wise_Saliency_of_Black-Box_Image_Classification_Models_ACCV_2020_paper.pdf
    """

    def __init__(
        self,
        k: int,
        p1: float = 0.0,
    ) -> None:
        """
        :param k: int
            Number of colors to used during perturbation.
        :param p1: float
            Debias probability, typically paired with the same probability used in mask generation.

        :raises: ValueError
            If p1 not in in [0, 1].
        :raises: ValueError
            If k < 1.
        """
        if p1 < 0 or p1 > 1:
            raise ValueError(f"Input p1 value of {p1} is not within the expected [0,1] range.")
        self.p1 = p1

        if k < 1:
            raise ValueError(f"Input k value of {k} is not within the expected >0 range.")
        self.k = k

    @override
    def generate(
        self,
        image_conf: np.ndarray,
        perturbed_conf: np.ndarray,
        perturbed_masks: np.ndarray,
    ) -> np.ndarray:
        """
        Warning: this implementation returns a different shape than typically expected by
        this interface. Instead of `[nClasses x H x W]`, saliency maps of shape
        `[kColors x nClasses x H x W]` are generated, one per color per class.

        :param image_conf: np.ndarray
            Reference image predicted class-confidence vector, as a
            `numpy.ndarray`, for all classes that require saliency map
            generation.
            This should have a shape `[nClasses]`, be float-typed and with
            values in the [0,1] range.
        :param perturbed_conf: np.ndarray
            Perturbed image predicted class confidence matrix.
            Classes represented in this matrix should be congruent to classes
            represented in the `image_conf` vector.
            This should have a shape `[nMasks x nClasses]`, be float-typed and
            with values in the [0,1] range.
        :param perturbed_masks: np.ndarray
            Perturbation masks `numpy.ndarray` over the reference image.
            This should be parallel in association to the classification
            results input into the `perturbed_conf` parameter.
            This should have a shape `[kColors x nMasks x H x W]`, and values in range
            [0, 1], where a value closer to 1 indicate areas of the image that
            are *unperturbed*.

        :return: np.ndarray
            Generated visual saliency heatmap for each input class as a
            float-type `numpy.ndarray` of shape `[kColors x nClasses x H x W]`.

        :raises: ValueError
            If number of perturbations masks and respective confidence lengths do not match.
        """
        if len(perturbed_conf) != perturbed_masks.shape[1]:
            raise ValueError("Number of perturbation masks and respective confidence lengths do not match.")

        sal_maps = []
        # Compute unmasked regions
        perturbed_masks = 1 - perturbed_masks
        m0 = 1 - perturbed_masks.sum(axis=0)
        for k_masks in perturbed_masks:
            # Debias based on the MC-RISE paper
            sal = weight_regions_by_scalar(
                scalar_vec=perturbed_conf,
                # Factoring out denominator from self.k * k_masks / self.p1 - m0 / (1 - self.p1)
                # to avoid divide by zero. Only acts as a normalization factor
                masks=self.k * k_masks * (1 - self.p1) - m0 * self.p1,
                inv_masks=False,
                normalize=False,
            )

            # Normalize final saliency map
            sal = maxabs_scale(sal.reshape(sal.shape[0], -1), axis=1).reshape(sal.shape)

            # Ensure saliency map in range [-1, 1]
            sal = np.clip(sal, -1, 1)

            sal_maps.append(sal)

        return np.asarray(sal_maps)

    @override
    def get_config(self) -> dict[str, Any]:
        return {
            "p1": self.p1,
            "k": self.k,
        }
