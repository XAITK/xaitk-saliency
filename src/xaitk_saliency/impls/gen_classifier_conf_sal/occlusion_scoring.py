"""Implementation of OcclusionScoring scorer"""

import numpy as np
from sklearn.preprocessing import maxabs_scale

from xaitk_saliency import GenerateClassifierConfidenceSaliency
from xaitk_saliency.utils.masking import weight_regions_by_scalar


class OcclusionScoring(GenerateClassifierConfidenceSaliency):
    """
    This saliency implementation transforms black-box
    image classification scores into saliency heatmaps. This should
    require a sequence of per-class confidences predicted on the
    reference image, a number of per-class confidences as predicted
    on perturbed images, as well as the masks of the reference image
    perturbations (as would be output from a `PerturbImage` implementation).

    The perturbation masks used by the following implementation are
    expected to be of type integer. Masks containing values of type
    float are rounded to the nearest value and binarized
    with value 1 replacing values greater than or equal to half of
    the maximum value in mask after rounding while 0 replaces the rest.
    """

    def generate(
        self,
        image_conf: np.ndarray,
        perturbed_conf: np.ndarray,
        perturbed_masks: np.ndarray,
    ) -> np.ndarray:
        """
        Generate saliency maps

        :param image_conf: np.ndarray
            Reference confidence lengths from the reference image
        :param perturbed_conf: np.ndarray
            Perturbed confidence lengths from the reference image
        :param perturbed_masks: np.ndarray
            Perturbation masks `numpy.ndarray` over the reference image.

        :return: np.ndarray
            Generated visual saliency heatmap.
        """
        if len(image_conf) != len(perturbed_conf[0]):
            raise ValueError("Number of classes in original image and perturbed image do not match.")

        if len(perturbed_conf) != len(perturbed_masks):
            raise ValueError("Number of perturbation masks and respective confidence lengths do not match.")

        # Iterating through each class confidence and compare it with
        # its perturbed twin
        diff = image_conf - perturbed_conf

        # Weighting perturbed regions with respective difference in confidence
        sal = weight_regions_by_scalar(diff, perturbed_masks)

        # Normalize final saliency map
        sal = maxabs_scale(sal.reshape(sal.shape[0], -1), axis=1).reshape(sal.shape)

        # Ensure saliency map in range [-1, 1]
        return np.clip(sal, -1, 1)

    def get_config(self) -> dict:
        """
        Get the configuration dictionary of the OcclusionScoring instance.

        Returns:
            dict[str, Any]: Configuration dictionary.
        """
        return {}
