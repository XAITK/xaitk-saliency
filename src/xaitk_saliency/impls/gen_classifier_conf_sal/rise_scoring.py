from typing import Any

import numpy as np
from sklearn.preprocessing import maxabs_scale

from xaitk_saliency import GenerateClassifierConfidenceSaliency
from xaitk_saliency.utils.masking import weight_regions_by_scalar


class RISEScoring(GenerateClassifierConfidenceSaliency):
    """
    Saliency map generation based on the original RISE implementation.
    This version utilizes only the input perturbed image confidence predictions
    and does not utilize reference image confidences.
    This implementation also takes influence from debiased RISE and may take an
    optional debias probability, `p1` (0 by default).
    In the original paper this is paired with the same probability used in RISE
    perturbation mask generation (see the `p1` parameter in
    :class:`xaitk_saliency.impls.perturb_image.rise.RISEGrid`).

    Based on Hatakeyama et. al:
    https://openaccess.thecvf.com/content/ACCV2020/papers/Hatakeyama_Visualizing_Color-wise_Saliency_of_Black-Box_Image_Classification_Models_ACCV_2020_paper.pdf
    """

    def __init__(
        self,
        p1: float = 0.0,
    ) -> None:
        """
        Generate RISE-based saliency maps with optional p1 de-biasing.

        :param p1: De-biasing parameter based on the masking probability.
            This should be a float value in the [0, 1] range.

        :raises ValueError: Input `p1` was not in the [0,1] range.
        """
        if p1 < 0 or p1 > 1:
            raise ValueError(f"Input p1 value of {p1} is not within the expected [0,1] range.")
        self.p1 = p1

    def generate(
        self,
        image_conf: np.ndarray,
        perturbed_conf: np.ndarray,
        perturbed_masks: np.ndarray,
    ) -> np.ndarray:
        if len(perturbed_conf) != len(perturbed_masks):
            raise ValueError("Number of perturbation masks and respective confidence lengths do not match.")

        # The RISE method does not use the difference of confidences, but just
        # the perturbed image confidences. The reference confidences are not
        # used here.

        # Weighting perturbed regions with respective difference in confidence
        sal = weight_regions_by_scalar(perturbed_conf, perturbed_masks - self.p1, inv_masks=False, normalize=False)

        # Normalize final saliency map
        sal = maxabs_scale(sal.reshape(sal.shape[0], -1), axis=1).reshape(sal.shape)

        # Ensure saliency map in range [-1, 1]
        return np.clip(sal, -1, 1)

    def get_config(self) -> dict[str, Any]:
        return {
            "p1": self.p1,
        }
