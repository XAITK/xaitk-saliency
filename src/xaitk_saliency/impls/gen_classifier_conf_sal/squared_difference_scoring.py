import numpy as np

from xaitk_saliency import GenerateClassifierConfidenceSaliency
from xaitk_saliency.utils.masking import weight_regions_by_scalar


class SquaredDifferenceScoring(GenerateClassifierConfidenceSaliency):
    """
    This saliency implementation transforms black-box confidence predictions
    from a classification-style network into saliency heatmaps. This should
    require a sequence of classification scores predicted on the reference
    image, a number of classification scores predicted on perturbed images, as
    well as the masks of the reference image perturbations (as would be output
    from a `PerturbImage` implementation).

    This implementation uses the squared difference of the reference scores and
    the perturbed scores to compute the saliency maps. This gives an indication
    of general saliency without distinguishing between positive and negative.
    The resulting maps are normalized between the range [0,1].

    Based on Greydanus et. al:
    https://arxiv.org/abs/1711.00138
    """

    def generate(self, reference: np.ndarray, perturbed: np.ndarray, perturbed_masks: np.ndarray) -> np.ndarray:
        if len(reference) != len(perturbed[0]):
            raise ValueError("Number of classes in original image and perturbed image do not match.")

        if len(perturbed) != len(perturbed_masks):
            raise ValueError("Number of perturbation masks and respective confidence lengths do not match.")

        # Based on equations 2 and 4 from Greydanus et al., '18
        diff = 0.5 * ((reference - perturbed) ** 2).sum(axis=1, keepdims=True)
        sal = weight_regions_by_scalar(diff, perturbed_masks)[0]

        # Normalize saliency map to [0,1]
        sal -= sal.min()
        return sal / sal.max()

    def get_config(self) -> dict:
        return {}
