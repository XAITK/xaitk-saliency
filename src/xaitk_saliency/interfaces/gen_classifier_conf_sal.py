import abc

import numpy as np
from smqtk_core import Plugfigurable


class GenerateClassifierConfidenceSaliency(Plugfigurable):
    """
    Visual saliency map generation interface whose implementations transform
    black-box image classification scores into saliency heatmaps.

    This should require a sequence of per-class confidences predicted on the
    reference image, a number of per-class confidences as predicted on
    perturbed images, as well as the masks of the reference image perturbations
    (as would be output from a
    :class:`xaitk_saliency.interfaces.perturb_image.PerturbImage`
    implementation).

    Implementations should use this input to generate a visual saliency
    heatmap for each input “class” in the input.
    This is both an effort to vectorize the operation for optimal performance,
    as well as to allow some algorithms to take advantage of differences in
    classification behavior for other classes to influence heatmap generation.
    For classifiers that generate many class label predictions, it is intended
    that only a subset of relevant class predictions need be provided here if
    computational performance is a consideration.
    """

    @abc.abstractmethod
    def generate(
        self,
        image_conf: np.ndarray,
        perturbed_conf: np.ndarray,
        perturbed_masks: np.ndarray,
    ) -> np.ndarray:
        """
        Generate a visual saliency heatmap matrix given the black-box
        classifier output on a reference image, the same classifier output on
        perturbed images and the masks of the visual perturbations.

        Perturbation mask input into the `perturbed_masks` parameter here is
        equivalent to the perturbation mask output from a
        :meth:`xaitk_saliency.interfaces.perturb_image.PerturbImage.perturb`
        method implementation.
        These should have the shape `[nMasks x H x W]`, and values in range
        [0, 1], where a value closer to 1 indicate areas of the image that
        are *unperturbed*.
        Note the type of values in masks can be either integer, floating point
        or boolean within the above range definition.
        Implementations are responsible for handling these expected variations.

        Generated saliency heatmap matrices should be floating-point typed and
        be composed of values in the [-1,1] range.
        Positive values of the saliency heatmaps indicate regions which increase
        class confidence scores, while negative values indicate regions which
        decrease class confidence scores according to the model that generated
        input confidence values.

        :param image_conf:
            Reference image predicted class-confidence vector, as a
            `numpy.ndarray`, for all classes that require saliency map
            generation.
            This should have a shape `[nClasses]`, be float-typed and with
            values in the [0,1] range.
        :param perturbed_conf:
            Perturbed image predicted class confidence matrix.
            Classes represented in this matrix should be congruent to classes
            represented in the `image_conf` vector.
            This should have a shape `[nMasks x nClasses]`, be float-typed and
            with values in the [0,1] range.
        :param perturbed_masks:
            Perturbation masks `numpy.ndarray` over the reference image.
            This should be parallel in association to the classification
            results input into the `perturbed_conf` parameter.
            This should have a shape `[nMasks x H x W]`, and values in range
            [0, 1], where a value closer to 1 indicate areas of the image that
            are *unperturbed*.

        :return: Generated visual saliency heatmap for each input class as a
            float-type `numpy.ndarray` of shape `[nClasses x H x W]`.
        """

    def __call__(
        self,
        image_conf: np.ndarray,
        perturbed_conf: np.ndarray,
        perturbed_masks: np.ndarray,
    ) -> np.ndarray:
        """Alias for :meth:`.GenerateClassifierConfidenceSaliency.generate`."""
        return self.generate(image_conf, perturbed_conf, perturbed_masks)
