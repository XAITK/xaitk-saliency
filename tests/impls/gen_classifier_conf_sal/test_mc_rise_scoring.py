import numpy as np
import pytest
from smqtk_core.configuration import configuration_test_helper
from syrupy.assertion import SnapshotAssertion

from tests import EXPECTED_MASKS_4x6
from tests.test_utils import CustomFloatSnapshotExtension
from xaitk_saliency.impls.gen_classifier_conf_sal.mc_rise_scoring import MCRISEScoring


@pytest.fixture
def snapshot_custom(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot.use_extension(lambda: CustomFloatSnapshotExtension())


class TestMCRiseScoring:
    def test_init_outofrange_config(self) -> None:
        """Test catching an out of range config value."""
        with pytest.raises(ValueError, match=r"Input p1 value of -0\.3 is not within the expected \[0,1\] range\."):
            MCRISEScoring(k=4, p1=-0.3)

        with pytest.raises(ValueError, match=r"Input p1 value of 5 is not within the expected \[0,1\] range\."):
            MCRISEScoring(k=1, p1=5)

        with pytest.raises(ValueError, match=r"Input k value of 0 is not within the expected >0 range\."):
            MCRISEScoring(k=0, p1=0.2)

    def test_configuration(self) -> None:
        """Test configuration aspects."""
        inst = MCRISEScoring(k=2, p1=0.747)
        for inst_i in configuration_test_helper(inst):
            assert np.allclose(inst_i.p1, 0.747)
            assert inst_i.k == 2

    def test_bad_alignment(self) -> None:
        """Test passing misaligned perturbed confidence vector and masks input."""
        test_confs = np.ones((3, 2))
        test_masks = np.ones((3, 4, 3, 3))

        inst = MCRISEScoring(k=3)

        with pytest.raises(
            ValueError,
            match=r"Number of perturbation masks and respective confidence lengths do not match",
        ):
            inst.generate(test_confs[0], test_confs, test_masks)

    def test_2class_scoring(self, snapshot_custom: SnapshotAssertion) -> None:
        """Test for expected output when given known input data."""
        test_ref_confs = np.array([0, 0.2])
        # Mock classification results for the test masked regions.
        test_pert_confs = np.array([[0.00, 0.33, 0.66, 0.33, 0.66, 1.00], [1.00, 0.66, 0.33, 0.66, 0.33, 0.00]]).T

        inst = MCRISEScoring(k=1)
        sal = inst.generate(test_ref_confs, test_pert_confs, np.asarray([EXPECTED_MASKS_4x6]))

        snapshot_custom.assert_match(sal)
