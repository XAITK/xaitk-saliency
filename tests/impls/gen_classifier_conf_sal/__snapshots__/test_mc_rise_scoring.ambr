# serializer version: 1
# name: TestMCRiseScoring.test_2class_scoring
  array([[[[0.  , 0.  , 0.33, 0.33, 0.66, 0.66],
           [0.  , 0.  , 0.33, 0.33, 0.66, 0.66],
           [0.33, 0.33, 0.66, 0.66, 1.  , 1.  ],
           [0.33, 0.33, 0.66, 0.66, 1.  , 1.  ]],

          [[1.  , 1.  , 0.66, 0.66, 0.33, 0.33],
           [1.  , 1.  , 0.66, 0.66, 0.33, 0.33],
           [0.66, 0.66, 0.33, 0.33, 0.  , 0.  ],
           [0.66, 0.66, 0.33, 0.33, 0.  , 0.  ]]]])
# ---
